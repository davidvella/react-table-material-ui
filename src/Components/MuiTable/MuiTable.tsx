import React, { FC } from 'react';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import {
  useTable,
  useSortBy,
  UseSortByColumnProps,
  TableInstance,
  UseFiltersColumnProps,
  ColumnInstance,
  UseResizeColumnsHeaderProps,
  usePagination,
  useRowSelect,
  Column,
  CellProps
} from 'react-table'
import { makeStyles, createStyles, Theme, TableSortLabel, withStyles, Paper, Checkbox } from '@material-ui/core';
import { MuiTableProps } from './MuiTable.types'
import { MuiTablePagination } from '../MuiTablePagination/MuiTablePagination';

/**
 * TypeScript Interface needed to compile.
 */
interface TableColumn<D extends object = {}>
  extends ColumnInstance<D>,
  UseSortByColumnProps<D>,
  UseResizeColumnsHeaderProps<D>,
  UseFiltersColumnProps<D> { }

/**
 * Styles for the Table
 */
const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    visuallyHidden: {
      border: 0,
      clip: 'rect(0 0 0 0)',
      height: 1,
      margin: -1,
      overflow: 'hidden',
      padding: 0,
      position: 'absolute',
      top: 20,
      width: 1,
    },
    tableWrapper: {
      overflow: 'auto'
    },
    selectedRow: {
      background: '#c2dbff !important'
    }
  }),
);

/**
 * ALternate Row Colours for usability.
 */
const StyledTableRow = withStyles((theme: Theme) =>
  createStyles({
    root: {
      '&:nth-child(odd)': {
        backgroundColor: theme.palette.background.default
      }
    }
  }),
)(TableRow);

/**
 * 
 * @param param0 
 */
export const MuiTable: FC<MuiTableProps> = (props) => {
  // Load values from props.
  const {
    columnsDef,
    data,
    // Default is 30
    initialPageSize = 30,
    rowsPerPageOptions,
    serverSide = false,
    pageCount,
    onColumnSortChange,
    onChangeRowsPerPage,
    onChangePage,
    onRowsSelect,
    isRowSelectable = true
  } = props;

  // Add the selection row to the table.
  const columns = React.useMemo(
    () => columnSelectionDecorator(columnsDef, isRowSelectable),
    [columnsDef, isRowSelectable],
  );

  // Use the state and functions returned from useTable to build your UI
  const {
    getTableProps,
    getTableBodyProps,
    headerGroups,
    prepareRow,
    page, // Instead of using 'rows', we'll use page,
    // which has only the rows for the active page.
    // The rest of these things are super handy, too ;)
    canPreviousPage,
    canNextPage,
    nextPage,
    previousPage,
    setPageSize,
    selectedFlatRows,
    state: { pageSize, pageIndex, sortBy, selectedRowPaths },
  } = useTable<object>(
    {
      columns,
      data,
      initialState: { pageSize: initialPageSize, pageIndex: 0 }, // Pass our hoisted table state
      manualPagination: serverSide, // Tell the usePagination
      // hook that we'll handle our own data fetching
      // This means we'll also have to provide our own
      // pageCount.
      pageCount: pageCount,
      manualSorting: serverSide,// Tell the useSortBy
      // hook that we'll sort our own data 
      // This means we'll also have to provide our own
      // pageCount.
    },
    useSortBy,
    usePagination,
    useRowSelect
  ) as TableInstance<object>;
  const classes = useStyles();

  // Listen for changes in column state and use the state to fetch new data
  React.useEffect(() => {
    if (onColumnSortChange) {
      onColumnSortChange(sortBy)
    }
  }, [onColumnSortChange, sortBy])

  // Listen for changes in pagination and use the state to fetch our new data
  React.useEffect(() => {
    if (onChangePage) {
      onChangePage(pageIndex)
    }
  }, [onChangePage, pageIndex])

  // Listen for changes in pageSize and use the state to fetch our new data
  React.useEffect(() => {
    if (onChangeRowsPerPage) {
      onChangeRowsPerPage(pageSize)
    }
  }, [onChangeRowsPerPage, pageSize])

  // Listen for changes in row selection and send data back to the user.
  React.useEffect(() => {
    if (onRowsSelect) {
      onRowsSelect(selectedFlatRows);
    }
  }, [onRowsSelect, selectedFlatRows, selectedRowPaths])

  // Render the UI for your table
  return (
    <Paper>
      <div className={classes.tableWrapper}>
        <Table {...getTableProps()} stickyHeader>
          <TableHead>
            {headerGroups.map(headerGroup => (
              <TableRow {...headerGroup.getHeaderGroupProps()}>
                {headerGroup.headers.map(c => {
                  const column = (c as unknown) as TableColumn<object>
                  return (
                    <TableCell {...column.getHeaderProps()}>
                      <TableSortLabel
                        active={column.isSorted}
                        direction={column.isSortedDesc ? "desc" : "asc"}
                        hideSortIcon={column.disableSortBy}
                        {...column.getSortByToggleProps()}
                      >
                        {column.render('Header')}
                        {column.isSorted ? (
                          <span className={classes.visuallyHidden}>
                            {column.isSortedDesc ? 'sorted descending' : 'sorted ascending'}
                          </span>
                        ) : null}
                      </TableSortLabel>
                    </TableCell>
                  )
                })}
              </TableRow>
            ))}
          </TableHead>
          <TableBody {...getTableBodyProps()}>
            {page.map(
              (row) => {
                prepareRow(row);
                return (
                  <StyledTableRow {...row.getRowProps()} className={row.isSelected ? classes.selectedRow : ''} hover={true}>
                    {row.cells.map(cell => {
                      return (
                        <TableCell {...cell.getCellProps()} tabIndex={0}>
                          {cell.render('Cell')}
                        </TableCell>
                      )
                    })}
                  </StyledTableRow>
                )
              }
            )}
          </TableBody>
        </Table>
      </div>
      <MuiTablePagination
        rowsPerPageOptions={rowsPerPageOptions}
        canNextPage={canNextPage}
        canPreviousPage={canPreviousPage}
        onClickNext={() => nextPage()}
        onClickPrevious={() => previousPage()}
        rowsPerPage={pageSize}
        onChangeRowsPerPage={e => {
          setPageSize(Number(e.target.value))
        }}
      />
    </Paper>


  )
}

/**
 * Adds the selection row if selected
 * @param columns 
 * @param isRowSelectable 
 */
const columnSelectionDecorator = (columns: Column<object>[], isRowSelectable: boolean) => {
  if (isRowSelectable) {
    columns.unshift(selectionColumn);
  }
  return columns;
}

// Let's make a column for selection
const selectionColumn: Column<object> =
{
  id: 'selection',
  // The header can use the table's getToggleAllRowsSelectedProps method
  // to render a checkbox
  Header: ({ getToggleAllRowsSelectedProps, isAllRowsSelected }: CellProps<any>) => (<div>
    <Checkbox {...getToggleAllRowsSelectedProps()} indeterminate={isAllRowsSelected} />
  </div>),
  // The cell can use the individual row's getToggleRowSelectedProps method
  // to the render a checkbox
  Cell: ({ row }: CellProps<any>) => (
    <div>
      <Checkbox {...row.getToggleRowSelectedProps()} />
    </div>
  ),
  disableSortBy: true
}