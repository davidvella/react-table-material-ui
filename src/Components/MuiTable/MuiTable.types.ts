import { Column, Row, SortingRule } from "react-table";

export interface MuiTableProps {
  /**
   * Columns used to describe table. Must be either an array of objects describing a column
   */
  columns: Column<any>[];
  /**
   * Data used to describe table. Must be either an array containing objects of key/value pairs with values that are strings or numbers, or arrays of strings or numbers (Ex: data: [{"Name": "Joe", "Job Title": "Plumber", "Age": 30}, {"Name": "Jane", "Job Title": "Electrician", "Age": 45}] or data: [["Joe", "Plumber", 30], ["Jane", "Electrician", 45]])
   */
  data: any[];
  /**
   * Number of rows allowed per page
   */
  initialPageSize?: number;
  /**
   * Customizes the options of the rows per page select field. If less than two options are available, no select field will be displayed.
   */
  rowsPerPageOptions?: Array<number>;
  /**
   * Enable remote data source
   */
  serverSide?: boolean;
  /**
   * User provided override for total number of pages
   */
  pageCount?: number;
  /**
   * Callback function to get data if serverSide is true.
   */
  serverSideFetchData?:(
    pageIndex: number,
    pageSize: number,
  ) => void;
  /**
   * Callback function to sort data if serverSide is true.
   */
  serverSideSort?: (
    sortBy: Array<SortingRule<any>>
  ) => void

}
