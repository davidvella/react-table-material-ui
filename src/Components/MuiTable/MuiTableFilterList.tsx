import { FC } from "react";
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';
import React from "react";
import { Typography, Button, GridList } from "@material-ui/core";

/**
 * Styles for the FilterList
 */
const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        root: {
            padding: '24px 24px 36px 24px',
            fontFamily: 'Roboto',
          },
          header: {
            flex: '0 0 auto',
            marginBottom: '16px',
            width: '100%',
            display: 'flex',
            justifyContent: 'space-between',
          },
          title: {
            display: 'inline-block',
            marginLeft: '7px',
            color: theme.palette.text.primary,
            fontSize: '14px',
            fontWeight: 500,
          },
          noMargin: {
            marginLeft: '0px',
          },
          reset: {
            alignSelf: 'left',
          },
          resetLink: {
            marginLeft: '16px',
            fontSize: '12px',
            cursor: 'pointer',
          },
          filtersSelected: {
            alignSelf: 'right',
          },
          checked: {},
          gridListTile: {
            marginTop: '16px',
          },
    }),
);


export const MuiTableFilterList: FC<any> = (props) => {
    const classes = useStyles();

    return(
        <div className={classes.root}>
        <div className={classes.header}>
          <div className={classes.reset}>
            <Typography
              variant="body2"
              className={classes.title}>
              FILTERS
            </Typography>
            <Button
              color="primary"
              className={classes.resetLink}
              tabIndex={0}
              aria-label={"RESET"}
              data-testid={'filterReset-button'}>
              RESET
            </Button>
          </div>
          <div className={classes.filtersSelected} />
        </div>
        <GridList cellHeight="auto" cols={2} spacing={34}>

        </GridList>
      </div>
    )
}