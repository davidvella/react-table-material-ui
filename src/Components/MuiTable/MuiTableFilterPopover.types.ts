export interface MuiTableFilterPopoverProps {
  /**
   * This is the DOM element, or a function that returns the DOM element, that may be used to set the position of the popover.
   */
  anchorEl?: null | Element | ((element: Element) => Element);
  /**
   * Callback function to set the anchor el.
   */
  updateAnchorEl?: (anchorEl: null | Element) => void;
}
