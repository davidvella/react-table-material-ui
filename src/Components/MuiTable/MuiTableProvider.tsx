import React, { createContext, FC, useMemo, useReducer } from "react";
import { IMuiTableProviderProps, IMuiTableContext, MuiTableAction } from "./MuiTableProvider.types";
import { MuiTableReducer } from "./MuiTableReducer";
import { OptionTypeBase, ValueType } from "react-select";

/**
 * The Initial state for the context.
 */
const initialState: IMuiTableContext = { filterValues: new Map<string, OptionTypeBase[] | null>(), columns: [] };

/**
 * The initial context
 */
export const MuiTableContext = createContext<{ state: IMuiTableContext, dispatch: (action: MuiTableAction) => void }>({ state: initialState, dispatch: (action) => {} });

/**
 * Context Provider
 * @param props 
 */
export const MuiTableProvider: FC<IMuiTableProviderProps> = (props) => {

  /**
   * Values from the props
   */
  const {
    columns,
    children
  } = props;

  /**
   * Populate dictionary with filter values
   */
  const defaultDictionary = useMemo(() => {
    const dict = new Map<string,  ValueType<OptionTypeBase> | undefined>();

    columns.filter(col => !col.disableFilters).forEach(col => {
      dict.set(col.id, null);
    })

    return dict
  }, [columns]);

  // new
  const [state, dispatch] = useReducer(MuiTableReducer, {filterValues:defaultDictionary, columns:columns});

  return (
    <MuiTableContext.Provider
      value={{
        state: state,
        dispatch: dispatch
      }}
    >
      {children}
    </MuiTableContext.Provider>
  );
};
