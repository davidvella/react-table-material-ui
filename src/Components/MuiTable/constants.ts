export const SELECT_COL_ID = 'selection'