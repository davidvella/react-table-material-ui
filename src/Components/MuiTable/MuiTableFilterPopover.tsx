import React, { FC, useContext } from "react";
import { Popover, Typography, Button } from "@material-ui/core";
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';
import { MuiTableFilterPopoverProps } from "./MuiTableFilterPopover.types";
import { MuiTableFilterList } from "./MuiTableFilterList";
import { MuiTableContext } from "./MuiTableProvider";

/**
 * Styles for the Table
 */
const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        root: {
            display: 'flex',
            justifyContent: 'left',
            flexWrap: 'wrap',
            margin: '0px 16px 0px 16px',
            padding: '24px 24px 36px 24px',
            fontFamily: 'Roboto',
        },
        paper: {
            overflowX: "unset!important" as any,
            overflowY: "unset!important" as any,
            maxWidth: 600
        },
        header: {
            flex: '0 0 auto',
            marginBottom: '16px',
            width: '100%',
            display: 'flex',
            justifyContent: 'space-between',
        },
        title: {
            display: 'inline-block',
            color: theme.palette.text.primary,
            fontSize: '14px',
            fontWeight: 500,
        },
        noMargin: {
            marginLeft: '0px',
        },
        reset: {
            alignSelf: 'left',
        },
        resetLink: {
            marginLeft: '16px',
            fontSize: '12px',
            cursor: 'pointer',
        },
        filtersSelected: {
            alignSelf: 'right',
        },
        checked: {},
        gridListTile: {
            marginTop: '16px',
        },
    }),
);

/**
 * A Popover used to display filter values content on top of the table. 
 * @param param0 
 */
export const MuiTableFilterPopover: FC<MuiTableFilterPopoverProps> = (props) => {
    // Map Context to columns
    const { dispatch } = useContext(MuiTableContext);
    const classes = useStyles();
    return (
        <Popover
            classes={{ paper: classes.paper }}
            open={Boolean(props.anchorEl)}
            anchorEl={props.anchorEl}
            elevation={2}
            onClose={() => {
                if (props.updateAnchorEl) {
                    props.updateAnchorEl(null);
                }
            }}
            anchorOrigin={{
                vertical: 'bottom',
                horizontal: 'center',
            }}
            transformOrigin={{
                vertical: 'top',
                horizontal: 'center',
            }}
        >
            <div className={classes.root}>
                <div className={classes.header}>
                    <div className={classes.reset}>
                        <Typography
                            variant="body2"
                            className={classes.title}>
                            FILTERS
                        </Typography>
                        <Button
                            color="primary"
                            className={classes.resetLink}
                            tabIndex={0}
                            aria-label={"RESET"}
                            data-testid={'filterReset-button'}
                            onClick={() => dispatch({ type: "resetFilter" })}>
                            RESET
                        </Button>
                    </div>
                    <div className={classes.filtersSelected} />
                </div>
                <MuiTableFilterList className={classes.root} />
            </div>

        </Popover>
    );
}