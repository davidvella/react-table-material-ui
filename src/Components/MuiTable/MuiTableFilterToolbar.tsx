import { FC, useContext, HTMLAttributes, useMemo } from "react";
import React from "react";
import { Theme, createStyles, Chip } from "@material-ui/core";
import { MuiTableContext } from "./MuiTableProvider";
import makeStyles from "@material-ui/styles/makeStyles";
import { uniqueId } from "lodash";

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        root: {
            display: 'flex',
            justifyContent: 'left',
            flexWrap: 'wrap',
            margin: '0px 16px 0px 16px',
        },
        chip: {
            margin: '8px 8px 0px 0px',
        },
    }),
);


/**
 * Render all filters based on context columns
 * @param props 
 */
export const MuiTableFilterToolbar: FC<HTMLAttributes<HTMLDivElement>> = () => {

    // styles
    const classes = useStyles();

    // Map Context to columns
    const { state: {
        filterValues
    } } = useContext(MuiTableContext);

    // Get values of all filter columns.
    const flatFilterValues = useMemo(() => Array.from(filterValues.values()), [filterValues])

    const filterChipSingleValue = (index: number, item: any) => {
        return (
            <Chip
                label={item.label}
                key={uniqueId(`${index}`)}
                className={classes.chip}
            />
        );
    };

    return (
        <div className={classes.root}>
            {flatFilterValues.map((item: any, index: number) => {
                if (item != null) {
                    if (Array.isArray(item)) {
                        return item.map((filterItem, filterItemIndex) =>
                            filterChipSingleValue(filterItemIndex, filterItem)
                        )
                    } else {
                        return filterChipSingleValue(index, item);
                    }
                }
                else {
                    return (null);
                }

            })}
        </div>
    )
}