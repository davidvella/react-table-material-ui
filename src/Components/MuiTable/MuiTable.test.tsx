import React, { FC } from 'react'
import { render, cleanup, fireEvent } from '@testing-library/react'
import { MuiTable } from './MuiTable';
// this adds custom jest matchers from jest-dom
import "@testing-library/jest-dom/extend-expect";
import "snapshot-diff";

const data = [
    {
        firstName: 'tanner',
        lastName: 'linsley',
        age: 29,
        visits: 100,
        status: 'In Relationship',
        progress: 50,
    },
    {
        firstName: 'derek',
        lastName: 'perkins',
        age: 40,
        visits: 40,
        status: 'Single',
        progress: 80,
    },
    {
        firstName: 'joe',
        lastName: 'bergevin',
        age: 45,
        visits: 20,
        status: 'Complicated',
        progress: 10,
    },
]

const App: FC<any> = (props) => {
    const columns = React.useMemo(
        () => [
            {
                Header: 'Name',
                columns: [
                    {
                        Header: 'First Name',
                        accessor: 'firstName',
                    },
                    {
                        Header: 'Last Name',
                        accessor: 'lastName',
                    },
                ],
            },
            {
                Header: 'Info',
                columns: [
                    {
                        Header: 'Age',
                        accessor: 'age',
                    },
                    {
                        Header: 'Visits',
                        accessor: 'visits',
                    },
                    {
                        Header: 'Status',
                        accessor: 'status',
                    },
                    {
                        Header: 'Profile Progress',
                        accessor: 'progress',
                    },
                ],
            },
        ],
        []
    )

    return <MuiTable columns={columns} data={data} initialPageSize={50} />
}

describe('<MuiTable />', () => {

    afterEach(cleanup);

    it('should renders a basic table', () => {
        const { getByText, asFragment } = render(<App />)

        expect(getByText('tanner')).toBeInTheDocument()
        expect(getByText('linsley')).toBeInTheDocument()
        expect(getByText('29')).toBeInTheDocument()
        expect(getByText('100')).toBeInTheDocument()
        expect(getByText('In Relationship')).toBeInTheDocument()

        expect(asFragment()).toMatchSnapshot()
    });

});