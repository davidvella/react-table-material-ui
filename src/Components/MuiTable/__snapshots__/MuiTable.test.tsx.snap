// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MuiTable /> should renders a basic table 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="makeStyles-tableWrapper-2"
    >
      <table
        class="MuiTable-root"
      >
        <thead
          class="MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head"
              colspan="2"
              scope="col"
            >
              <span
                aria-disabled="false"
                class="MuiButtonBase-root MuiTableSortLabel-root"
                hidden=""
                role="button"
                tabindex="0"
                title="Toggle SortBy"
              >
                Name
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                  />
                </svg>
              </span>
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head"
              colspan="4"
              scope="col"
            >
              <span
                aria-disabled="false"
                class="MuiButtonBase-root MuiTableSortLabel-root"
                hidden=""
                role="button"
                tabindex="0"
                title="Toggle SortBy"
              >
                Info
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                  />
                </svg>
              </span>
            </th>
          </tr>
          <tr
            class="MuiTableRow-root MuiTableRow-head"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head"
              colspan="1"
              scope="col"
            >
              <span
                aria-disabled="false"
                class="MuiButtonBase-root MuiTableSortLabel-root"
                role="button"
                style="cursor: pointer;"
                tabindex="0"
                title="Toggle SortBy"
              >
                First Name
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                  />
                </svg>
              </span>
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head"
              colspan="1"
              scope="col"
            >
              <span
                aria-disabled="false"
                class="MuiButtonBase-root MuiTableSortLabel-root"
                role="button"
                style="cursor: pointer;"
                tabindex="0"
                title="Toggle SortBy"
              >
                Last Name
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                  />
                </svg>
              </span>
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head"
              colspan="1"
              scope="col"
            >
              <span
                aria-disabled="false"
                class="MuiButtonBase-root MuiTableSortLabel-root"
                role="button"
                style="cursor: pointer;"
                tabindex="0"
                title="Toggle SortBy"
              >
                Age
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                  />
                </svg>
              </span>
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head"
              colspan="1"
              scope="col"
            >
              <span
                aria-disabled="false"
                class="MuiButtonBase-root MuiTableSortLabel-root"
                role="button"
                style="cursor: pointer;"
                tabindex="0"
                title="Toggle SortBy"
              >
                Visits
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                  />
                </svg>
              </span>
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head"
              colspan="1"
              scope="col"
            >
              <span
                aria-disabled="false"
                class="MuiButtonBase-root MuiTableSortLabel-root"
                role="button"
                style="cursor: pointer;"
                tabindex="0"
                title="Toggle SortBy"
              >
                Status
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                  />
                </svg>
              </span>
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head"
              colspan="1"
              scope="col"
            >
              <span
                aria-disabled="false"
                class="MuiButtonBase-root MuiTableSortLabel-root"
                role="button"
                style="cursor: pointer;"
                tabindex="0"
                title="Toggle SortBy"
              >
                Profile Progress
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                  />
                </svg>
              </span>
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root WithStyles(ForwardRef(TableRow))-root-68"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body"
              tabindex="0"
            >
              tanner
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body"
              tabindex="0"
            >
              linsley
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body"
              tabindex="0"
            >
              29
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body"
              tabindex="0"
            >
              100
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body"
              tabindex="0"
            >
              In Relationship
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body"
              tabindex="0"
            >
              50
            </td>
          </tr>
          <tr
            class="MuiTableRow-root WithStyles(ForwardRef(TableRow))-root-68"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body"
              tabindex="0"
            >
              derek
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body"
              tabindex="0"
            >
              perkins
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body"
              tabindex="0"
            >
              40
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body"
              tabindex="0"
            >
              40
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body"
              tabindex="0"
            >
              Single
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body"
              tabindex="0"
            >
              80
            </td>
          </tr>
          <tr
            class="MuiTableRow-root WithStyles(ForwardRef(TableRow))-root-68"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body"
              tabindex="0"
            >
              joe
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body"
              tabindex="0"
            >
              bergevin
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body"
              tabindex="0"
            >
              45
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body"
              tabindex="0"
            >
              20
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body"
              tabindex="0"
            >
              Complicated
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body"
              tabindex="0"
            >
              10
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="MuiToolbar-root MuiToolbar-regular makeStyles-root-69 makeStyles-toolbar-70 MuiToolbar-gutters"
    >
      <div
        class="makeStyles-spacer-71"
      />
      <p
        class="MuiTypography-root makeStyles-caption-72 MuiTypography-body2 MuiTypography-colorInherit"
      >
        Rows per page:
      </p>
      <div
        class="MuiInputBase-root makeStyles-input-76 makeStyles-selectRoot-73"
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby=" "
          class="MuiSelect-root MuiSelect-select makeStyles-select-74 MuiSelect-selectMenu MuiInputBase-input MuiInputBase-inputSelect"
          role="button"
          tabindex="0"
        >
          50
        </div>
        <input
          type="hidden"
          value="50"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSelect-icon makeStyles-selectIcon-75"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
      <div
        class="makeStyles-actions-78"
      >
        <button
          aria-label="previous page"
          class="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 16.59L10.83 12l4.58-4.59L14 6l-6 6 6 6 1.41-1.41z"
              />
            </svg>
          </span>
        </button>
        <button
          aria-label="next page"
          class="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.59 16.59L13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
